### PR TITLE
[air] Example to run tuning and analyze results

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This is a notebook showing how to tune an xgboost model and analyze the results.

Also adds a `get_dataframe()` method to `ResultsGrid` to fetch the trial results.

Depends on #24483 for toctree.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 1]
```diff
@@ -0,0 +1,2497 @@
...
+    "from ray.tune.tune_config import TuneConfig\n",
+    "from ray.tune.tuner import Tuner"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a93b242c",
+   "metadata": {},
+   "source": [
+    "We'll define a utility function to create a Ray dataset from the Sklearn dataset. We expect the target column to be in the dataframe, so we'll add it to the dataframe manually.\n",
```
Comment (Yard1) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:51]:
```suggestion
    "We'll define a utility function to create a Ray Dataset from the Sklearn dataset. We expect the target column to be in the dataframe, so we'll add it to the dataframe manually.\n",
```

**Review by Yard1 — COMMENTED** [review]
This looks great. I think it would benefit from reducing the verbosity of the tuner, the output is far too long.

**Comment by krfricke** [comment]
I don't think it matters that much as it will be rendered in a notebook, but I've changed to default verbosity (2) now.

File: python/ray/tune/result_grid.py [hunk 2]
```diff
@@ -80,6 +81,36 @@ def get_best_result(
...
+        """Return dataframe of all trials with their configs and reported results.
+
+        Per default, this returns the last reported results for each trial.
+
+        If ``filter_metric`` and ``filter_mode`` are set, the results from each
+        trial are filtered for this metric and mode. For example, if
+        ``filter_metric="some_metric"`` and ``filter_mode="max"``, for each trial,
+        every received result is checked, and the one where ``some_metric`` is
+        maximal is returned.
+
```
Comment (xwjiang2010) [inline python/ray/tune/result_grid.py:98]: Can we provide the schema of the df in doc string?

File: python/ray/tune/result_grid.py [hunk 3]
```diff
@@ -80,6 +81,36 @@ def get_best_result(
             )
         )
 
+    def get_dataframe(
```
Comment (xwjiang2010) [inline python/ray/tune/result_grid.py:84]: Add test?

**Review by maxpumperla — APPROVED** [review]
This is great! I think we should really think about exposing these examples on our docs soon, given that AIR is still deeply hidden. So, the "risk" of too many users being exposed is fairly low and we can slowly but surely build up a nice example gallery for official release.

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 4]
```diff
@@ -0,0 +1,918 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "c62191af",
+   "metadata": {},
+   "source": [
+    "# Analyzing results from hyperparameter tuning\n",
+    "In this example we'll run a hyperparameter tuning experiment and analyze the results.\n",
```
Comment (amogkam) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:9]: Nit: maybe add a bit more description here.

In this example, we will go through how you can use Ray AIR to run a distributed hyperparameter experiment to find optimal hyperparameters for an XGBoost model.

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 5]
```diff
@@ -0,0 +1,918 @@
...
+    "from ray.tune.tuner import Tuner"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a93b242c",
+   "metadata": {},
+   "source": [
+    "We'll define a utility function to create a Ray Dataset from the Sklearn dataset. We expect the target column to be in the dataframe, so we'll add it to the dataframe manually.\n",
+    "We don't do any test/validation splitting here - pelase check out our other examples for this."
```
Comment (amogkam) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:52]:
```suggestion
    "We don't do any test/validation splitting here - please check out our other examples for this."
```

Comment (amogkam) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:52]: Which examples?

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 6]
```diff
@@ -0,0 +1,918 @@
...
+   "source": [
+    "print(train_dataset)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1d996bdd",
+   "metadata": {},
+   "source": [
+    "Ray AIR allows us to define and tune preprocessors. For instance, we may want to decide if we should scale the \"Wilderness area\" parameters or the \"Soil type\" parameters. For both, we'll just try standard scalers:"
```
Comment (amogkam) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:111]: Maybe add a sentence on what preprocessors are and why I would use them? Also link to the actual docs for more information on preprocessors.

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 7]
```diff
@@ -0,0 +1,918 @@
...
+    "# XGBoost specific params\n",
+    "params = {\n",
+    "    \"tree_method\": \"approx\",\n",
+    "    \"objective\": \"multi:softmax\",\n",
+    "    \"eval_metric\": [\"mlogloss\", \"merror\"],\n",
+    "    \"num_class\": 8,\n",
+    "}\n",
+    "\n",
+    "trainer = XGBoostTrainer(\n",
+    "    scaling_config={\"num_workers\": 2},\n",
```
Comment (amogkam) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:152]: Add a comment on what scaling config is?

Until this point, I actually didn't know that each trial was itself going to be distributed training. Might be worth mentioning this earlier.

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 8]
```diff
@@ -0,0 +1,918 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tuner = Tuner(\n",
+    "    trainer,\n",
+    "    param_space={\n",
+    "        \"params\": {\"max_depth\": tune.randint(2, 8)},\n",
+    "        \"preprocessor\": tune.grid_search([prep_v1, prep_v2]),\n",
+    "    },\n",
+    "    tune_config=TuneConfig(num_samples=4, metric=\"train-mlogloss\", mode=\"min\"),\n",
```
Comment (amogkam) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:183]: Do the metric and mode matter if we are not using a searcher or scheduler? 

Maybe we can just add a basic searcher to the example.

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 9]
```diff
@@ -0,0 +1,918 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "ab642705",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
+       "== Status ==<br>Current time: 2022-05-09 17:29:41 (running for 00:04:05.52)<br>Memory usage on this node: 9.3/16.0 GiB<br>Using FIFO scheduling algorithm.<br>Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/4.63 GiB heap, 0.0/2.0 GiB objects<br>Current best trial: 4544b_00000 with train-mlogloss=0.559286 and parameters={'params': {'max_depth': 7}, 'preprocessor': StandardScaler(columns=['Wilderness_Area_0', 'Wilderness_Area_1', 'Wilderness_Area_2', 'Wilderness_Area_3'], ddof=0, stats={'mean(Wilderness_Area_0)': 0.4488650836815763, 'mean(Wilderness_Area_1)': 0.051434393781884025, 'mean(Wilderness_Area_2)': 0.4360736094951567, 'mean(Wilderness_Area_3)': 0.06362691304138296, 'std(Wilderness_Area_0)': 0.4973783472700716, 'std(Wilderness_Area_1)': 0.22088208826922595, 'std(Wilderness_Area_2)': 0.4958965785292558, 'std(Wilderness_Area_3)': 0.2440871339874498})}<br>Result logdir: /Users/kai/ray_results/XGBoostTrainer_2022-05-09_17-25-34<br>Number of trials: 8/8 (8 TERMINATED)<br><table>\n",
```
Comment (amogkam) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:204]: Let's decrease the verbosity for the example? I think @Yard1 is enabling this by default right?

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 10]
```diff
@@ -0,0 +1,918 @@
...
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Preprocessor 1 minimum log loss: 0.559286\n",
+      "Preprocessor 2 minimum log loss: 0.559286\n",
+      "Preprocessor 1 minimum error: 0.197125\n",
+      "Preprocessor 2 minimum error: 0.197125\n"
```
Comment (amogkam) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:828]: Seems like choice of preprocessor doesn't make a difference?

Comment (amogkam) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:828]: Should we add a sentence explaining the takeaway (so as a user I know how to interpret these results)?

**Review by amogkam — COMMENTED** [review]
Looks good @krfricke!

Left some comments in the example. I think we should also try to link to other parts of the docs throughout the example on where users can find more information (preprocessors, XGBoostTrainer, Tuner/ray tune)

**Comment by amogkam** [comment]
Could we also show the necessary pip installs to run the example? @bveeramani had this in his examples- I think it's a great point.

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 9 above)
Comment (Yard1) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:204]: Verbosity will be set to 1 by default (simple trial results)

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 5 above)
Comment (krfricke) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:52]: Happy to link once they are actually integrated in the docs. This is a bit of a chicken egg problem

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 10 above)
Comment (krfricke) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:828]: Yeah actually I'll have to either leave out preprocessing here or find something else that actually makes sense. Decision trees don't care about monotonic transformations...

File: python/ray/tune/result_grid.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/result_grid.py:98]: I've added some info in the `Returns` part, but we can't easily provide the full schema as it will depend on the reported metrics (including possibly changing auto-generated metrics).

File: python/ray/tune/result_grid.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/result_grid.py:84]: Functionality is tested in `test_experiment_analysis.py` and we only pass through here. We can do that but maybe should separate this out from this PR.

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 5 above)
Comment (krfricke) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:52]: I just removed the comment for now, it doesn't provide anything in the current example

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 8 above)
Comment (krfricke) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:183]: There's a comment explaining them, and we should keep them here as it's good practice if they want to switch to non-random search later

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 10 above)
Comment (Yard1) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb]: encoding categorical columns, maybe?

Comment (krfricke) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb]: The dataset is pretty much perfect in terms of preprocessing. Let's do dataset and preprocessor tuning somewhere else.
We should probably also focus on showing just one thing per example, and in this case this would be the analysis of the results dataframe.

**Comment by krfricke** [comment]
Yes, good point - added!

**Review by xwjiang2010 — APPROVED** [review]

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 11]
```diff
@@ -0,0 +1,918 @@
...
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "2022-05-09 17:25:30,768\tINFO services.py:1484 -- View the Ray dashboard at \u001B[1m\u001B[32mhttp://127.0.0.1:8265\u001B[39m\u001B[22m\n"
+     ]
+    }
+   ],
+   "source": [
+    "def get_training_data() -> ray.data.Dataset:\n",
```
Comment (xwjiang2010) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:215]: This won't work with `tuner.resume` though. Do we want the example to showcase that?

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 8 above)
Comment (xwjiang2010) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb]: I am wondering if it makes sense to echo back to the code above and point out why the metric is `train-mlogloss`

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb [hunk 12]
```diff
@@ -0,0 +1,918 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "# This will fetch the best result according to the `metric` and `mode` specified\n",
+    "# in the `TuneConfig` above:\n",
+    "\n",
+    "best_result = results.get_best_result()\n",
+    "\n",
+    "print(\"Best result error rate\", best_result.metrics[\"train-merror\"])"
```
Comment (xwjiang2010) [inline doc/source/ray-air/examples/analyze_tuning_results.ipynb:909]: do we also want to show `mlogloss`?

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 8 above)
Comment (xwjiang2010) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb]: Also do we want to briefly mention that Tune's optimization is based on single metric although xgboost reports two to Tune?

File: python/ray/tune/result_grid.py (hunk 3 above)
Comment (xwjiang2010) [reply python/ray/tune/result_grid.py]: Sounds good!

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 11 above)
Comment (krfricke) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:215]: It should with `lambda:` right? I think it's fine to leave out, this example focuses on analysis, not on the problem

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 8 above)
Comment (krfricke) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb]: Good points, added

File: doc/source/ray-air/examples/analyze_tuning_results.ipynb (hunk 12 above)
Comment (krfricke) [reply doc/source/ray-air/examples/analyze_tuning_results.ipynb:909]: Added

**Review by maxpumperla — APPROVED** [review]
